x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/config_panel.tsx

```diff
@@ -174,6 +187,24 @@ export function LayerPanels(
             }
           }}
           onRemoveLayer={() => {
+            const datasourcePublicAPI = props.framePublicAPI.datasourceLayers?.[layerId];
+            const datasourceId = datasourcePublicAPI?.datasourceId;
+            const layerDatasourceState = datasourceStates?.[datasourceId]
+              ?.state as IndexPatternPrivateState;
+
+            const trigger = props.uiActions.getTrigger(UPDATE_FILTER_REFERENCES_TRIGGER);
+            const action = props.uiActions.getAction(UPDATE_FILTER_REFERENCES_ACTION);
+
+            action?.execute({
+              trigger,
+              toDataView: null,
+              fromDataView: layerDatasourceState.layers[layerId].indexPatternId,
```

Comment:
The editor frame shouldn't have any knowledge about the shape of the datasource / visualization state. It should only pass it around. Both accessing `layers` as well as `layers.indexPatternId` break this assumption. If we require this information here, add a new getter to the `Datasource` definition, implement it for the datasource and call it here.

Reply:
@flash1293  could you please review again?